### ISSUE TITLE
OwinContext.GetUserManager returning null

### ISSUE THREAD

**Issue by terryclancy** [Shazwazza/UmbracoIdentity#46]
Hi,

I finally got UmbracoIdentity with Facebook login working fine in a test project :-)  And I have now created a fresh new Umbraco 7.4.1 installation where I cam combining all my project code with my Umbraco Identity code.

I am now getting an "Object reference not set to an instance of an object" error in my IdentitySurfaceController when trying to use "UserManager" where OwinContext.GetUserManager<UmbracoMembersUserManager<UmbracoApplicationMember>>();  is returning null.  This means that I am getting to the line flagged with // ERROR !!!!!  below

```

        private UmbracoMembersUserManager<Models.TT_Identity.UmbracoApplicationMember> _userManager;


        protected IOwinContext OwinContext
        {

            get {
                var tempOwinContext = Request.GetOwinContext();
                return tempOwinContext;
            }
        }

        public UmbracoMembersUserManager<UmbracoApplicationMember> UserManager
        {
            get
            {
                // return _userManager ?? (_userManager = OwinContext
                // .GetUserManager<UmbracoMembersUserManager<UmbracoApplicationMember>>());
                if  (_userManager == null)
                {
                    _userManager = OwinContext.GetUserManager<UmbracoMembersUserManager<UmbracoApplicationMember>>();
                }
                if (_userManager == null)
                {
                    // ERROR !!!!!
                    var i = 0;
                    i = i++;
                }
                return _userManager;
            }
        }


```
Note that UmbracoIdentityStartup.cs is running which included the following code

            //Single method to configure the Identity user manager for use with Umbraco
            app.ConfigureUserManagerForUmbracoMembers<UmbracoApplicationMember>();

Can I assume that the UserManagerForUmbracoMembers is not being initialized properly ? Perhaps I forgot to do something in my new project ? I am not familiar with Owin yet so any assistance in what is wrong or how to approach understanding and debugging this issue would be appreciated.

Thanks very much

Terry Clancy

ClanceZ

**Comment by Shazwazza** [comment]
Can you post your whole startup class?

**Comment by terryclancy** [comment]
Hi Shannon,

Thanks - here you go:


```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using Microsoft.AspNet.Identity;
using Microsoft.AspNet.Identity.Owin;
using Microsoft.Owin;
using Microsoft.Owin.Security.Cookies;
using UmbracoIdentity;
using TeraTastic.Models.UmbracoIdentity;
using TeraTastic;
using Owin;
using Umbraco.Web;

[assembly: OwinStartup("UmbracoIdentityStartup", typeof(UmbracoIdentityStartup))]

namespace TeraTastic
{

    /// <summary>
    /// OWIN Startup class for UmbracoIdentity 
    /// </summary>
    public class UmbracoIdentityStartup : UmbracoDefaultOwinStartup
    {
        public override void Configuration(IAppBuilder app)
        {
            base.Configuration(app);

            //Single method to configure the Identity user manager for use with Umbraco
            app.ConfigureUserManagerForUmbracoMembers<UmbracoApplicationMember>();



            // Enable the application to use a cookie to store information for the 
            // signed in user and to use a cookie to temporarily store information 
            // about a user logging in with a third party login provider 
            // Configure the sign in cookie
            app.UseCookieAuthentication(
                        //You can modify these options for any customizations you'd like
                        new FrontEndCookieAuthenticationOptions());

            app.UseExternalSignInCookie(DefaultAuthenticationTypes.ExternalCookie);

            // Uncomment the following lines to enable logging in with third party login providers

            //app.UseMicrosoftAccountAuthentication(
            //  clientId: "",
            //  clientSecret: "");

            //app.UseTwitterAuthentication(
            //  consumerKey: "",
            //  consumerSecret: "");

            //app.UseFacebookAuthentication(
            //  appId: "",
            //  appSecret: "");

            app.UseFacebookAuthentication(
            appId: "53736854???????",
            appSecret: "d5d919da18265aabfba86e??????????");

            //app.UseGoogleAuthentication(
            //  clientId: "",
            //  clientSecret: ""); 



        }


    }
}
```

Terry Clancy

ClanceZ

**Comment by terryclancy** [comment]
Hi Shannon,

To debug this issue I have removed the reference to UmbracoIdentity.dll in my project, downloaded the source code for UmbracoIdentity and added the UmbracoIdentity  project to my solution so that I can now debug and step into UmbracoIdentity code. With that done I added a breakpoint in UmbracoIdentityStartup.cs.  I was then able to determine that in my UmbracoIdentityStartup.cs the call to 

app.ConfigureUserManagerForUmbracoMembers<UmbracoApplicationMember>();

above steps into 

UmbracoIdentity/ExternalLoginStore.cs 

at the following method call

public static void ConfigureUserManagerForUmbracoMembers<T>(this IAppBuilder app,
            ApplicationContext appContext = null,
            IdentityEnabledMembersMembershipProvider membershipProvider = null)
            where T : UmbracoIdentityMember, new()

I stepped through that method and did not see any  obvious problems but I am not sure what to look for.

Now that I have the source code for UmbracoIdentity in my project, is there anything you can suggest as an approach to help debug this issue ?

Terry Clancy

ClanceZ

**Comment by terryclancy** [comment]
Hi again Shannon,

OK I have some more info that may help.  it appears that at startup the following method in AppBuilderExtensions.cs   never gets called.  Should this get called at startup ?

public static void ConfigureUserManagerForUmbracoMembers<TManager, TUser>(this IAppBuilder app,
            ApplicationContext appContext,
            Func<IdentityFactoryOptions<TManager>, IOwinContext, TManager> userManager)
            where TManager : UmbracoMembersUserManager<TUser>
            where TUser : UmbracoIdentityMember, new()
        {
            if (appContext == null) throw new ArgumentNullException("appContext");
            if (userManager == null) throw new ArgumentNullException("userManager");

            //Don't proceed if the app is not ready
            if (appContext.IsConfigured == false
                || appContext.DatabaseContext == null
                || appContext.DatabaseContext.IsDatabaseConfigured == false) return;

            //Configure Umbraco user manager to be created per request
          **  app.CreatePerOwinContext<TManager>(userManager);**
        }

Thanks

Terry Clancy

ClanceZ

**Comment by terryclancy** [comment]
Shannon,

OK I have finally solved the problem.

The issue was that I have left your sample source in my solution including /Models/UmbracoIdentity/UmbracoApplicationMember.cs  however all my code uses my own version of this including all the files in /Models/TT_Identity including /Models/TT_Identity/UmbracoApplicationMember.cs.  The error was caused by the fact that in UmbracoIdentityStartup.cs I had not changed

`using TeraTastic.Models.UmbracoIdentity;`

to 

`using TeraTastic.Models.TT_Identity;`

and so UmbracoIdentityStartup.cs was using a different set of models than the rest of my application.  Replacing using TeraTastic.Models.UmbracoIdentity; with using TeraTastic.Models.TT_Identity; resolved the issue.

Thanks

Terry Clancy

ClanceZ
